jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/ColorRefinementIsomorphismInspector.java

```diff
@@ -135,17 +139,15 @@ public boolean isomorphismExists() {
             return false;
         }
 
-        ColorRefinementAlgorithm<V, E> colorRefinementAlgorithm1 = new ColorRefinementAlgorithm<>(graph1);
-        ColorRefinementAlgorithm<V, E> colorRefinementAlgorithm2 = new ColorRefinementAlgorithm<>(graph2);
+        Graph<Pair<V, Integer>, Pair<E, Integer>> graph = new UnionGraph<>(graph1, graph2);
```

Comment:
Use `AsGraphUnion`. The assertion that it copies the vertex sets is not true.

Moreover, I do not think that you really need the `Pair<V, Integer>`. I assume that you use the integer as a flag to distinguish between the two graphs. You can query the original graphs in order to figure out if a vertex belongs to the first or the second.

Reply:
If we should use `AGraphUnion`, we have to check, in which graph the vertex lies. Furthermore, we have to distinguish the incoming edges by the graphs. That is, we have to modify the color refinement algorithm for exactly this case.

If this should be done, the the following questions arise:
* How do I get the edges of a specific graph without checking if the edge is in a specific graph? Or how do I do this efficiently? `AsUnionGraph` mixes up the edges of both graphs.
* Should we implement two versions of the CR algorithm, one for the specific usage of isomorphism inspectation and one for general usage?
* We already implemented an individualization refinement algorithm based on this CR implementation. Should we implement two version, too: One for isomorphism checking and one for general usage?

Reply:
The union is disjoint, so you can distinguish edges the same way that you distinguish vertices. Given a vertex v you can query the vertex set of graph1 (expected constant time due to the hashtable) and figure out whether it belongs to graph1 or graph2.

You should be able to implement this without the PairSet or the UnionGraph inner classes.

I am having trouble understanding the last 2 bullets.

Reply:
We need to distiguish both graphs within the color refinement algorithm, because in there we call vertexSet() and incomingEdges(). We cannot do this from the outside (here: the isomorphism inspector). That is, we need to implement a different version of the current CR algorithm because it cannot distinguish two vertices/edges, which are equal but in both graphs.

Thus, we may implement one version, in which we use the union graph, and a second version, which is the current implementation.